File: java/src/main/java/com/fleetmgr/sdk/client/state/pilot/Connected.java
package com.fleetmgr.sdk.client.state.pilot;

import com.fleetmgr.interfaces.Channel;
import com.fleetmgr.interfaces.Role;
import com.fleetmgr.sdk.client.event.input.Event;
import com.fleetmgr.sdk.client.event.input.connection.ConnectionEvent;
import com.fleetmgr.sdk.client.event.input.user.UserEvent;
import com.fleetmgr.sdk.client.event.output.facade.OperationStarted;
import com.fleetmgr.sdk.client.state.pilot.connected.ValidatingChannels;
import com.fleetmgr.sdk.system.machine.BaseState;

import java.util.Collection;
import java.util.Optional;
import java.util.logging.Logger;

/**
 * Created by: Bartosz Nawrot
 * Date: 23.09.2018
 * Description:
 */
public class Connected extends BaseState {

    private static final Logger logger = Logger.getLogger(Connected.class.getName());

    private BaseState internalState;

    private Role initialRole;

    Connected(BaseState state, Role role, Collection<Channel> channels) {
        super(state);
        this.internalState = new ValidatingChannels(this, role, channels);
        this.initialRole = role;
    }

    @Override
    public void start() {
        internalState.start();
        backend.getHeartbeatHandler().start();
        listener.onEvent(new OperationStarted(initialRole));
        initialRole = null;
    }

    @Override
    public Optional<BaseState> onUserEvent(UserEvent event) {
        return handle(event);
    }

    @Override
    public Optional<BaseState> onConnectionEvent(ConnectionEvent event) {
        return handle(event);
    }

    private Optional<BaseState> handle(Event event) {
        Optional<BaseState> newState = internalState.handleEvent(event);
        boolean wasRecovering = internalState instanceof com.fleetmgr.sdk.client.state.device.connected.Recovering;

        while (newState.isPresent()) {
            logger.info("Connected transition: " + toString() + " -> Connected." + newState.toString());
            internalState = newState.get();

            internalState.start();
            newState = internalState.getInitialState();
        }

        if (internalState instanceof com.fleetmgr.sdk.client.state.device.connected.Released) {
            return Optional.of(new Disconnecting(this, wasRecovering));
        }

        return Optional.empty();
    }

    @Override
    public String toString() {
        return "Connected." + internalState;
    }
}

File: cpp/source/state/pilot/Connected.cpp
#include "state/pilot/Connected.hpp"

#include "state/pilot/connected/ValidatingChannels.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::state;
using namespace fm::state::pilot;

using namespace com::fleetmgr::interfaces;
using namespace com::fleetmgr::interfaces::facade::control;

using namespace event::input::user;
using namespace event::input::connection;
using namespace event::output;

Connected::Connected(IState& state, const std::vector<ChannelResponse>& openedChannels) :
    IState(state),
    internalState(std::make_unique<connected::ValidatingChannels>(*this, openedChannels))
{
}

std::unique_ptr<IState> Connected::start()
{
    backend.getHeartbeatHandler().start();
    internalState->start();
    listener.onEvent(std::make_shared<FacadeEvent>(FacadeEvent::OPERATION_STARTED));
    return nullptr;
}

std::unique_ptr<IState> Connected::handleUserEvent(const UserEvent& event)
{
    return onNewState(internalState->handleUserEvent(event));
}

std::unique_ptr<IState> Connected::handleConnectionEvent(const ConnectionEvent& event)
{
    return onNewState(internalState->handleConnectionEvent(event));
}

std::unique_ptr<IState> Connected::onNewState(std::unique_ptr<IState> newState)
{
    while (nullptr != newState.get())
    {
        listener.trace("Connected transition: " + internalState->toString() + " -> " + newState->toString());
        internalState.swap(newState);
        newState.reset(internalState->start().release());
    }
    return internalState->createOuterState();
}

std::string Connected::toString() const
{
    return "Connected::" + internalState->toString();
}
